cmd/clusterctl/internal/test/fake_objects.go

```diff
@@ -1371,3 +1377,136 @@ func FakeCRDList() []*apiextensionsv1.CustomResourceDefinition {
 		clusterInfrastructureIdentityCRD,
 	}
 }
+
+type FakeClusterClass struct {
+	namespace                                 string
+	name                                      string
+	infrastructureClusterTemplate             *unstructured.Unstructured
+	controlPlaneTemplate                      *unstructured.Unstructured
+	controlPlaneInfrastructureMachineTemplate *unstructured.Unstructured
+	workerMachineDeploymentClasses            []*FakeMachineDeploymentClass
+}
+
+func NewFakeClusterClass(namespace, name string) *FakeClusterClass {
+	return &FakeClusterClass{
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+func (f *FakeClusterClass) WithInfrastructureClusterTemplate(tmpl *unstructured.Unstructured) *FakeClusterClass {
+	f.infrastructureClusterTemplate = tmpl
+	return f
+}
+
+func (f *FakeClusterClass) WithControlPlaneTemplate(tmpl *unstructured.Unstructured) *FakeClusterClass {
+	f.controlPlaneTemplate = tmpl
+	return f
+}
+
+func (f *FakeClusterClass) WithControlPlaneInfrastructureTemplate(tmpl *unstructured.Unstructured) *FakeClusterClass {
+	f.controlPlaneInfrastructureMachineTemplate = tmpl
+	return f
+}
+
+func (f *FakeClusterClass) WithWorkerMachineDeploymentClasses(classes []*FakeMachineDeploymentClass) *FakeClusterClass {
+	f.workerMachineDeploymentClasses = classes
+	return f
+}
+
+func (f *FakeClusterClass) Objs() []client.Object {
+	// objMap map where the key is the object to which the owner reference to the cluster class should be added
+	// and the value dictates if the onwner ref needs to be added.
+	// This map also dual functions as a way to de-duplicate and template objects that are reused.
+	objMap := map[client.Object]bool{}
```

Comment:
Can a client.Object be used as a map key?

I wonder if it's comparable.

> As mentioned earlier, map keys may be of any type that is comparable. The language spec defines this precisely, but in short, comparable types are boolean, numeric, string, pointer, channel, and interface types, and structs or arrays that contain only those types. Notably absent from the list are slices, maps, and functions; these types cannot be compared using ==, and may not be used as map keys.
> https://go.dev/blog/maps#TOC_5.

Reply:
I assume it is safe as the only types which could be duplicate are Unstructured and Unstructured is essentially only nested maps?

I wonder what would happen with regular types (but we don't have to care about them here)

Reply:
Yup, since this is only for the test builder with just unstructured objects it works.